core/src/main/java/org/sonarsource/sonarlint/core/util/ProgressWrapper.java

```diff
@@ -19,22 +19,34 @@
  */
 package org.sonarsource.sonarlint.core.util;
 
+import javax.annotation.Nullable;
 import org.sonarsource.sonarlint.core.client.api.common.ProgressMonitor;
 import org.sonarsource.sonarlint.core.client.api.exceptions.CanceledException;
 
-import javax.annotation.Nullable;
-
 public class ProgressWrapper {
-  private ProgressMonitor handler;
 
-  public ProgressWrapper(@Nullable ProgressMonitor handler) {
+  private final ProgressMonitor handler;
+  private final float offset;
+  private final float factor;
+
+  private ProgressWrapper(float offset, float factor, @Nullable ProgressMonitor handler) {
+    this.offset = offset;
+    this.factor = factor;
     if (handler == null) {
       this.handler = new NoOpProgressMonitor();
     } else {
       this.handler = handler;
     }
   }
 
+  public ProgressWrapper(@Nullable ProgressMonitor handler) {
+    this(0.0f, 1.0f, handler);
+  }
+
+  public ProgressWrapper subProgress(float fromFraction, float toFraction) {
+    return new ProgressWrapper(offset + fromFraction * factor, (toFraction - fromFraction) * factor, handler);
```

Comment:
Indeed, but what if there are more than 1 sub level? Also, I wonder if we should not support updating the progress without changing the message.